File: 2023/12.rb
input = File.read(File.basename(__FILE__).gsub('rb', 'input')).strip.split("\n")

$cache = {}
def arrangements(str, template)
  return $cache[str + template.to_s] if $cache[str + template.to_s]

  if str.size == 0
    $cache[str + template.to_s] = template.size.zero? ? 1 : 0
  elsif str[0] == '.'
    $cache[str + template.to_s] = arrangements(str[1..], template)
  elsif str[0] == '?'
    $cache[str + template.to_s] = arrangements(str.sub('?', '.'), template) +
                                  arrangements(str.sub('?', '#'), template)
  elsif str[0] == '#'
    if template.size == 0 || str.size < template[0] || str[0...template[0]].chars.any? { _1 == '.' }
      0
    elsif template.size > 1
      return 0 if str.size < template[0] + 1 || str[template[0]] == '#'

      $cache[str + template.to_s] = arrangements(str[template[0] + 1..], template[1..])
    else
      $cache[str + template.to_s] = arrangements(str[template[0]..], template[1..])
    end
  end

  ### nice and slow
  # if str.index('?').nil?
  #   return str.scan(/#+/).map(&:size) == template ? 1 : 0
  # end

  # [arrangements(str.sub('?', '.'), template), arrangements(str.sub('?', '#'), template)].sum
end

p(input.map do |line|
  arrangements(line.split(' ')[0], line.split(' ')[1].split(',').map(&:to_i))
end.sum)

p(input.map do |line|
  arrangements(([line.split(' ')[0]] * 5).join('?'), line.split(' ')[1].split(',').map(&:to_i) * 5)
end.sum)

File: 2023/12.py
input = open(f"{__file__.replace('py', 'input')}").read().strip().split("\n")


cache = {}


def arrangements(string, template):
    if string + str(template) in cache:
        return cache[string + str(template)]
    if len(string) == 0:
        cache[string + str(template)] = 1 if len(template) == 0 else 0
        return cache[string + str(template)]
    if string.startswith("."):
        cache[string + str(template)] = arrangements(string.strip("."), template)
        return cache[string + str(template)]
    if string.startswith("?"):
        cache[string + str(template)] = arrangements(
            string.replace("?", ".", 1), template
        ) + arrangements(string.replace("?", "#", 1), template)
        return cache[string + str(template)]
    if string.startswith("#"):
        if (
            len(template) == 0
            or len(string) < template[0]
            or any(c == "." for c in string[0 : template[0]])
        ):
            return 0
        if len(template) > 1:
            if len(string) < template[0] + 1 or string[template[0]] == "#":
                return 0
            cache[string + str(template)] = arrangements(
                string[template[0] + 1 :], template[1:]
            )
            return cache[string + str(template)]
        else:
            cache[string + str(template)] = arrangements(
                string[template[0] :], template[1:]
            )
            return cache[string + str(template)]


total_sum = 0
for line in input:
    string, template = line.split(" ")
    template = [int(x) for x in template.split(",")]
    total_sum += arrangements(string, template)
print(total_sum)

total_sum = 0
for line in input:
    string, template = line.split(" ")
    template = [int(x) for x in template.split(",")]

    string = "?".join([string] * 5)
    template = template * 5
    total_sum += arrangements(string, template)
print(total_sum)
